mm/alignment.py

```diff
@@ -413,3 +417,261 @@ def run(self):
+
+    def run(self) -> None:
+        discarded_seqs = 0
+        counted_seqs = 0
+
+        start_differences = Counter()
+        end_differences = Counter()
+        differences = Counter()
+
+        hyp_alignments = rasr_cache.open_file_archive(self.hyp_alignment_cache.get())
+        hyp_alignments.setAllophones(self.hyp_allophone_file.get())
+        if isinstance(hyp_alignments, rasr_cache.FileArchiveBundle):
+            hyp_allophone_map = next(iter(hyp_alignments.archives.values())).allophones
+        else:
+            hyp_allophone_map = hyp_alignments.allophones
+
+        ref_alignments = rasr_cache.open_file_archive(self.ref_alignment_cache.get())
+        ref_alignments.setAllophones(self.ref_allophone_file.get())
+        if isinstance(ref_alignments, rasr_cache.FileArchiveBundle):
+            ref_allophone_map = next(iter(ref_alignments.archives.values())).allophones
+        else:
+            ref_allophone_map = ref_alignments.allophones
+
+        file_list = [tag for tag in hyp_alignments.file_list() if not tag.endswith(".attribs")]
+
+        for idx, hyp_seq_tag in enumerate(file_list, start=1):
+            hyp_word_starts, hyp_word_ends, hyp_seq_length = self._compute_word_boundaries(
+                hyp_alignments,
+                hyp_allophone_map,
+                hyp_seq_tag,
+                self.hyp_silence_phone,
+                self.hyp_upsample_factor,
+            )
+            assert len(hyp_word_starts) == len(
+                hyp_word_ends
+            ), f"Found different number of word starts ({len(hyp_word_starts)}) than word ends ({len(hyp_word_ends)}). Something seems to be broken."
+
+            if self.hyp_seq_tag_transform is not None:
+                ref_seq_tag = self.hyp_seq_tag_transform(hyp_seq_tag)
+            else:
+                ref_seq_tag = hyp_seq_tag
+
+            ref_word_starts, ref_word_ends, ref_seq_length = self._compute_word_boundaries(
+                ref_alignments,
+                ref_allophone_map,
+                ref_seq_tag,
+                self.ref_silence_phone,
+                self.ref_upsample_factor,
+            )
+            assert len(ref_word_starts) == len(
+                ref_word_ends
+            ), f"Found different number of word starts ({len(hyp_word_starts)}) than word ends ({len(hyp_word_ends)}) in reference. Something seems to be broken."
+
+            if len(hyp_word_starts) != len(ref_word_starts):
+                logging.warning(
+                    f"Sequence {hyp_seq_tag} ({idx} / {len(file_list)}:\n    Discarded because the number of words in alignment ({len(hyp_word_starts)}) does not equal the number of words in reference ({len(ref_word_starts)})."
+                )
+                discarded_seqs += 1
+                continue
+
+            # Sometimes different feature extraction or subsampling may produce mismatched lengths that are different by one frame, so cut off at the shorter length
```

Comment:
I just noticed that the comment explicitly states "one frame". Should we check that here? So adding `if hyp_seq_len - 1 <= ref_seq_len <= hyp_seq_len + 1` do what's below. I'm not sure how you considered this originally, maybe this check isn't needed.

Reply:
Thinking about this, it can actually be more than one frame once the subsampling factor comes into play. For example if one sequence has length 14 with subsampling 1 then a model with subsampling 4 would give an alignment of length 3 or 4 depending on floor/ceil division which would be upsampled to 12 or 16. So I wouldn't check a specific maximum frame difference.